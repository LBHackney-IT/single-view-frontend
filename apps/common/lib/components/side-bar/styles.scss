@import "../../../node_modules/lbh-frontend/lbh/base";
@import "../../../node_modules/govuk-frontend/govuk/helpers/visually-hidden";

.mtfh-sidebar {
  .lbh-summary-list {
    .govuk-summary-list__key,
    .govuk-summary-list__value {
      font-size: 1rem;
      font-weight: 400;
    }

    .govuk-summary-list__value {
      color: lbh-colour("lbh-grey-1");
    }
  }

  img {
    width: 100%;
  }

  .lbh-button {
    width: 100%;
  }

  .govuk-accordion__controls {
    @include govuk-visually-hidden-focusable();
  }
}
