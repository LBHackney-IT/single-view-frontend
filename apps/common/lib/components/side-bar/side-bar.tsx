import React, {
  Children,
  ReactElement,
  cloneElement,
  forwardRef,
  isValidElement,
} from "react";

import cn from "classnames";

import { useBreakpoint } from "../../hooks";
import { Accordion, AccordionItem, AccordionItemProps } from "../accordion";
import { Heading } from "../heading";

import type * as Polymorphic from "@radix-ui/react-polymorphic";
import "./styles.scss";

export interface SideBarSectionProps extends AccordionItemProps {
  isCollapsed?: boolean;
  heading?: string;
}

export type SideBarSectionComponent = Polymorphic.ForwardRefComponent<
  "div",
  SideBarSectionProps
>;

export const SideBarSection: SideBarSectionComponent = forwardRef(function SideBarSection(
  { children, heading, className, isCollapsed = false, ...props },
  ref,
) {
  if (isCollapsed) {
    return (
      <AccordionItem ref={ref} {...props}>
        {children}
      </AccordionItem>
    );
  }

  return (
    <div ref={ref} className={cn("mtfh-sidebar-section", className)} {...props}>
      {heading ? <Heading as="h2">{heading}</Heading> : undefined}
      {children}
    </div>
  );
});

export interface SideBarProps {
  id: string;
  top?: ReactElement;
  children:
    | ReactElement<SideBarSectionProps>
    | null
    | Array<ReactElement<SideBarSectionProps> | null>;
}

export type SideBarComponent = Polymorphic.ForwardRefComponent<"div", SideBarProps>;

export const SideBar: SideBarComponent = forwardRef(function SideBar(
  { as: SideBarComp = "div", id, top, children, className, ...props },
  ref,
) {
  const isDesktop = useBreakpoint("md");
  const sidebarClasses = cn("mtfh-sidebar", className);

  return (
    <SideBarComp ref={ref} className={sidebarClasses} {...props}>
      {top}
      {!isDesktop ? (
        <Accordion id={id}>
          {Children.map<
            ReactElement<SideBarSectionProps> | undefined,
            ReactElement<SideBarSectionProps> | null
          >(children, (child) =>
            child && isValidElement(child)
              ? cloneElement(child, {
                  isCollapsed: true,
                })
              : undefined,
          )}
        </Accordion>
      ) : (
        <div id={id}>{children}</div>
      )}
    </SideBarComp>
  );
});
