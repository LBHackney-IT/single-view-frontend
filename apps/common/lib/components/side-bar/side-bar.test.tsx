import React from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";
import { screen } from "@testing-library/react";

import { SideBar, SideBarSection } from "./side-bar";

test("it renders correctly on desktop", async () => {
  const { container } = render(
    <SideBar id="sidebar">
      <SideBarSection id="section-1" title="More Details">
        <div>Content</div>
      </SideBarSection>
    </SideBar>,
    { query: "lg" },
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders correctly on mobile", async () => {
  const { container } = render(
    <SideBar id="sidebar">
      <SideBarSection id="section-1" title="More Details">
        <div>Content</div>
      </SideBarSection>
    </SideBar>,
    { query: "base" },
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders a heading on desktop", async () => {
  const { container } = render(
    <SideBar id="sidebar">
      <SideBarSection id="section-1" title="More Details" heading="A Heading">
        <div>Content</div>
      </SideBarSection>
    </SideBar>,
    { query: "lg" },
  );
  expect(screen.getByRole("heading")).toHaveTextContent("A Heading");
  await testA11y(container);
});

test("it does not renders a heading on mobile", async () => {
  const { container } = render(
    <SideBar id="sidebar">
      <SideBarSection id="section-1" title="More Details" heading="A Heading">
        <div>Content</div>
      </SideBarSection>
    </SideBar>,
    { query: "base" },
  );
  expect(screen.queryByText("A Heading")).toBe(null);
  await testA11y(container);
});

test("accepts and ignores a null child", () => {
  const { container } = render(
    <SideBar id="sidebar">
      <SideBarSection id="section-1" title="More Details" heading="A Heading">
        <div>Content</div>
      </SideBarSection>
      {null}
      <SideBarSection id="section-2" title="More Details 2" heading="A Heading 2">
        <div>Content 2</div>
      </SideBarSection>
      {null}
    </SideBar>,
    { query: "base" },
  );
  expect(container).toMatchSnapshot();
});
