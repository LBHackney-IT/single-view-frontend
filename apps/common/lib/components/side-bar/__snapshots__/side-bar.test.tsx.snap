// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts and ignores a null child 1`] = `
<div>
  <div
    class="mtfh-sidebar"
  >
    <div
      class="govuk-accordion lbh-accordion"
      data-attribute="value"
      id="sidebar"
    >
      <div
        class="govuk-accordion__controls"
      >
        <button
          aria-expanded="false"
          class="govuk-accordion__open-all"
          type="button"
        >
          Open all
          <span
            class="govuk-visually-hidden"
          >
             sections
          </span>
        </button>
      </div>
      <div
        class="govuk-accordion__section"
      >
        <div
          class="govuk-accordion__section-header"
        >
          <h3
            class="lbh-heading-h5 govuk-accordion__section-heading"
          >
            <button
              aria-controls="sidebar-content-1"
              aria-expanded="false"
              class="govuk-accordion__section-button"
              id="accordion-heading-section-1"
              type="button"
            >
              More Details
              <span
                aria-hidden="true"
                class="govuk-accordion__icon"
              />
            </button>
          </h3>
        </div>
        <div
          aria-labelledby="accordion-heading-section-1"
          class="govuk-accordion__section-content"
          id="accordion-content-section-1"
        >
          <div>
            Content
          </div>
        </div>
      </div>
      <div
        class="govuk-accordion__section"
      >
        <div
          class="govuk-accordion__section-header"
        >
          <h3
            class="lbh-heading-h5 govuk-accordion__section-heading"
          >
            <button
              aria-controls="sidebar-content-2"
              aria-expanded="false"
              class="govuk-accordion__section-button"
              id="accordion-heading-section-2"
              type="button"
            >
              More Details 2
              <span
                aria-hidden="true"
                class="govuk-accordion__icon"
              />
            </button>
          </h3>
        </div>
        <div
          aria-labelledby="accordion-heading-section-2"
          class="govuk-accordion__section-content"
          id="accordion-content-section-2"
        >
          <div>
            Content 2
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`it renders correctly on desktop 1`] = `
<div>
  <div
    class="mtfh-sidebar"
  >
    <div
      id="sidebar"
    >
      <div
        class="mtfh-sidebar-section"
        id="section-1"
        title="More Details"
      >
        <div>
          Content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`it renders correctly on mobile 1`] = `
<div>
  <div
    class="mtfh-sidebar"
  >
    <div
      class="govuk-accordion lbh-accordion"
      data-attribute="value"
      id="sidebar"
    >
      <div
        class="govuk-accordion__controls"
      >
        <button
          aria-expanded="false"
          class="govuk-accordion__open-all"
          type="button"
        >
          Open all
          <span
            class="govuk-visually-hidden"
          >
             sections
          </span>
        </button>
      </div>
      <div
        class="govuk-accordion__section"
      >
        <div
          class="govuk-accordion__section-header"
        >
          <h3
            class="lbh-heading-h5 govuk-accordion__section-heading"
          >
            <button
              aria-controls="sidebar-content-1"
              aria-expanded="false"
              class="govuk-accordion__section-button"
              id="accordion-heading-section-1"
              type="button"
            >
              More Details
              <span
                aria-hidden="true"
                class="govuk-accordion__icon"
              />
            </button>
          </h3>
        </div>
        <div
          aria-labelledby="accordion-heading-section-1"
          class="govuk-accordion__section-content"
          id="accordion-content-section-1"
        >
          <div>
            Content
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
